packages/forklift-console-plugin/src/modules/Providers/views/details/components/DetailsSection/utils/getOpenstackProviderWebUILink.ts

```diff
@@ -0,0 +1,25 @@
+import { V1beta1Provider } from '@kubev2v/types';
+
+import { getProviderUIAnnotation } from './getProviderUIAnnotation';
+
+/**
+ * A function for auto calculating the OpenStack UI link.
+ * It extracts the provider's OpenStack UI link, E.g.  https://identity_service.com/dashboard
+ * rom the OpenStack URL, E.g. https://identity_service.com:5000/v3
+ */
+export const getOpenstackProviderWebUILink = (provider: V1beta1Provider): string => {
+  // Check for custom link
+  const customLink = getProviderUIAnnotation(provider);
+  if (customLink) {
+    return customLink;
+  }
+
+  // Calculate link using API URL
+  const newSuffix = '/dashboard';
+  const urlLinkRegexp = new RegExp('((:[0-9]+)?)/v3$', 'g');
+  const regexpResult = urlLinkRegexp.exec(provider?.spec?.url);
+
+  return provider?.spec?.url && regexpResult
+    ? provider?.spec?.url.slice(0, urlLinkRegexp.lastIndex - regexpResult[0].length) + newSuffix
+    : '';
```

Comment:
```suggestion
   const url = provider?.spec?.url || '';
   const suffix = '/v3';
   const newSuffix = '/dashboard';
   
   if (url.endsWith(suffix)) {
    return url.slice(0, -suffix.length) + newSuffix;
  } else {
    return undefined;
  }
```

my consern with current code is that it's not easy to understand, it uses regexp that is not imidiatly clear that it is doing endswith, it uses trinary `?` that is less clear then` if .. else` and it's using `provider?.spec?.url`  that has two optional `?`

Reply:
I know that using `if else` if preferable, but this case is a bit different than for previous vSphere provider:  we need to replace port (if exists) + "`v3`" path with `dashboard` path.
E.g. http://host:5000/v3 => http://host/dashboard.

Leaving the port won't always work since the dashboard is usually using the default port.

Using if else for that is more complicated...

Reply:
> Leaving the port won't always work since the dashboard is usually using the default port.

are you sure ? 

>  we need to replace port  + "v3"

do we only support v3 ?

I still have concerns about readability:

this regex does:
http://example.com:9696/v3 =>  http://example.com/dashboard
and
http://example.com:9696/mypath/v3 =>  http://example.com:9696/mypath/dashboard

do we conserve the port in the second case by mistake or by design ? when using regex it's not clear what was the intention of this code.

> Using if else for that is more complicated

when using if else ( !! in this case, regexp has it's uses !! ) we can be clear, correct and cover more use cases,
consider this example:

```ts
const url = provider?.spec?.url || '';
if (!url) {
  return undefined;
}

let urlObj = new URL(url)

# Replace port
urlObj.port = '';

# Check the path
if (urlObj.pathname.endsWith(/'v3')) {
   const newPathname = `${urlObj.patname.slice(0, -3)}/dashboard`;
   urlObj.patname = newPathname;
   
   return urlObj.toString();
else {
   return undefined;
}
```

this example is verbose, and uses a URL object that can cover more cases if needed